### PR TITLE
feat: add more comprehensive config validation checking

### PR DESCRIPTION
**Pull request by petebacondarwin**

### CODE REVIEW

File: packages/wrangler/src/config/environment.ts [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+   *
+   * NOTE: This field is not automatically inherited from the top level environment,
+   * and so must be specified in every named environment.
+   *
+   * @default `[]`
+   * @nonInheritable
+   */
+  r2_buckets: {
+    /** The binding name used to refer to the R2 bucket in the worker. */
+    binding: string;
```
Comment (Electroid) [inline packages/wrangler/src/config/environment.ts:220]: I think this should be `name`?

Comment (petebacondarwin) [reply packages/wrangler/src/config/environment.ts:220]: LOL! Sunil said that too.
This is currently what Wrangler1 does. So we are copying that here.
Happy to have a discussion about this moving forward.

Comment (petebacondarwin) [reply packages/wrangler/src/config/environment.ts:220]: Note that `kv_namespaces` also use `binding`.

Comment (petebacondarwin) [reply packages/wrangler/src/config/environment.ts:220]: But durable_objects use `name` 🤷

Comment (Electroid) [reply packages/wrangler/src/config/environment.ts:220]: Maybe we should move them all to `name`, with `binding` as deprecated? /cc @threepointone

Comment (threepointone) [reply packages/wrangler/src/config/environment.ts:220]: Yeah we should definitely do a sweep of all of these. Removing the unnecessary `bindings` key on durable, etc etc. I'm wary of doing the deprecation right now, this all would be so much simpler when we have auto toml editing.

**Review by threepointone — APPROVED** [review]
Me and Pete walked through this entire PR, and I annoyed him with my nits and a couple of followups, but it otherwise looks great to me. Tentatively stamping. If this breaks anything you can blame me since I said so.

**Comment by threepointone** [comment]
image

**Pull request merged**